File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/identifier/ClassIdentifier.java
package me.itzisonn_.meazy.parser.ast.expression.identifier;

import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.runtime.environment.Environment;
import me.itzisonn_.meazy.runtime.environment.EnvironmentUtils;
import org.jspecify.annotations.NullMarked;

@NullMarked
public class ClassIdentifier extends Identifier {
    public ClassIdentifier(String id) {
        super(id);
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, ProgramUnit parent) {
        throw new UnsupportedOperationException("Can't emit class identifier");
    }

    @Override
    public DataType getType(Environment environment, ProgramUnit parent) {
        return DataType.ofNonNull(EnvironmentUtils.resolveClassDesc(environment, id, false));
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/identifier/ClassIdentifier.kt
package me.itzisonn_.meazy.parser.ast.expression.identifier

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.DataType
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.runtime.environment.Environment
import me.itzisonn_.meazy.runtime.environment.EnvironmentUtils.resolveClassDesc

open class ClassIdentifier(id: String) : Identifier(id) {
    override fun emit(instructions: InstructionsSet, environment: Environment, parent: ProgramUnit) {
        error("Can't emit class identifier")
    }

    override fun getType(environment: Environment, parent: ProgramUnit): DataType {
        return DataType.ofNonNull(resolveClassDesc(environment, id, false))
    }
}